train.py

```diff
@@ -69,46 +70,88 @@ def get_args():
+                # refresh for the next packed sequence
+                temp_lengths.clear()
+                temp_ids.clear()
+            temp_ids.extend(input_ids)
+            temp_lengths.append(len(input_ids))
+        # add the last one
+        if temp_lengths:
+            packed_lengths.append(temp_lengths.copy())
+            packed_ids.append(temp_ids.copy())
+        # pad the packed_ids into max_window
+        padded_input_ids = np.zeros((len(packed_ids), max_window), dtype=np.int64)
+        padded_attention_masks = np.zeros((len(packed_ids), max_window), dtype=np.int64)
+        for i, example_input_ids in enumerate(packed_ids):
+            padded_input_ids[i, :len(example_input_ids)] = example_input_ids
+            # obtain the attention mask
+            attention_mask = create_flatten_attention_mask(packed_lengths[i], args.max_input_length)
+            padded_attention_masks[i] = attention_mask
+        model_inputs["input_ids"] = padded_input_ids
+        model_inputs["attention_mask"] = padded_attention_masks
+        # take the last valid eos token as the end of the sequence
+        first_eos_indices = [len(packed_ids[i]) - 1 for i in range(len(packed_ids))]
     else:
-        model_inputs = [inp + tar + tokenizer.eos_token for inp, tar in zip(inputs, targets)]
-   
-    model_inputs = tokenizer(model_inputs,
-                             max_length=args.max_input_length,
-                             padding="max_length",
-                             truncation=True)
-
-    assert tokenizer.eos_token_id == tokenizer.pad_token_id
-    # This relies on tokenizer.eos_token_id == tokenizer.pad_token_id
-    first_eos_indices = [
-        model_inputs["input_ids"][i].index(tokenizer.eos_token_id) 
-        if model_inputs["input_ids"][i][-1] == tokenizer.eos_token_id else args.max_input_length
-        for i in range(len(model_inputs["input_ids"]))
-    ]
+        assert tokenizer.eos_token_id == tokenizer.pad_token_id
+        model_inputs = tokenizer(model_inputs,
+                                 max_length=args.max_input_length,
+                                 # no padding to calculate the real length
+                                 padding=True,
+                                 truncation=True)
+        # This relies on tokenizer.eos_token_id == tokenizer.pad_token_id
+        first_eos_indices = [
+            model_inputs["input_ids"][i].index(tokenizer.eos_token_id)
+            if model_inputs["input_ids"][i][-1] == tokenizer.eos_token_id else args.max_input_length
+            for i in range(len(model_inputs["input_ids"]))
+        ]
 
     if args.compute_loss_on_input:
         assert args.compute_loss_on_instruction is False
         model_inputs["labels"] = [
-            inp[:first_eos_indices[i] + 1] + [-100] * (args.max_input_length - first_eos_indices[i] - 1)
+            list(inp[:first_eos_indices[i] + 1]) + [-100] * (args.max_input_length - first_eos_indices[i] - 1)
```

Comment:
Why was this necessary? Shouldn't it already be a list?

Reply:
In the case of `data_packing`, it is a part of the numpy array. So we should use list to convert it into a list.